### DISCUSSION TITLE
Are Minio and CouchDB Proxied URLs required?

### DISCUSSION THREAD

**Discussion by ikogan** [Budibase/budibase#8665] [Help]
The Budibase proxy seems to proxy `/` to the minio instance and `/db` to CouchDB. Are these required for Budibase to function or are they convenience URLs as part of compose/k8s?

I ask because I'm deploying CouchDB separately and using Ceph for S3 so it's getting a little challenging to make these proxies work correctly. I'm having a separate issue and want to try to eliminate `/whatever` being able to proxy to Ceph's ObjectStore. According to the architecture, it would seem that the browser _does_ access them through the proxy but it's not clear to me, based on the nginx config, how it can properly authorize access to CouchDB and S3 by _just_ going through the proxy unless the worker delivers the backend credentials (S3 access and secret keys) to the client browser.

**Comment by Rory-Powell** [comment]
Hi @ikogan, thanks for the detailed question. 

To answer these questions:
> - Are these required for Budibase to function or are they convenience URLs as part of compose/k8s?
>  - How it can properly authorize access to CouchDB and S3 by just going through the proxy unless the worker delivers the backend credentials?

The CouchDB route is exposed over the nginx proxy only for accessing the admin interface. i.e. `http://localhost:10000/db/_utils`, where the client will be prompted for credentials, it wouldn't break anything if that route was omitted. The server and worker access the CouchDB container directly, and use the configured username and password from environment variables. 

The MinIO route is exposed over the nginx proxy for reading files, and writes or updates to files in the buckets are done via direct container access in server and worker, similar to the database where environment variables will be used. This means that bucket reads are public to facilitate this, rather than routing file reads through our own api. There is a feature in progress that will introduce presigned urls for accessing files, to better secure reads. 

In summary:
- Using a separate CouchDB instance is supported
   - The proxy is not required. 
- Using a separate S3-like object store is supported
   - However in self host it's currently required that the object store be available on the default `/` route. 
   - Hence, your Ceph instance would need to be accessible behind the proxy. 
   - To make this possible you will likely need to wait for this one to be merged: https://github.com/Budibase/budibase/pull/7413
   - Additionally as a part of the presigned url changes that are being made, we can look into removing the requirement for a custom object store to be hosted behind the proxy
